crates/biome_js_analyze/src/lint/nursery/use_readonly_class_properties.rs

```diff
@@ -0,0 +1,449 @@
+    class_declaration: &JsClassDeclaration,
+    private_only: bool,
+) -> Option<Vec<TextRangeWithText>> {
+    let constructor = class_declaration
+        .members()
+        .iter()
+        .find_map(|member| match member {
+            AnyJsClassMember::JsConstructorClassMember(member) => Some(member),
+            _ => None,
+        })?;
+
+    let constructor_params = constructor.parameters().ok()?;
+    let params: Vec<TextRangeWithText> = constructor_params
+        .parameters()
+        .iter()
+        .filter_map(|param| match param.ok()? {
+            AnyJsConstructorParameter::TsPropertyParameter(ts_property) => Some(ts_property),
+            _ => None,
+        })
+        .filter(|param| {
+            let is_mutable = param
+                .modifiers()
+                .iter()
+                .find_map(|x| TsReadonlyModifier::cast(x.into_syntax()))
+                .is_none();
+
+            let is_private = param.modifiers().iter().any(|x| {
+                x.as_ts_accessibility_modifier()
+                    .is_some_and(|modifier| modifier.is_private())
+            });
+
+            is_mutable && (!private_only || is_private)
+        })
+        .filter_map(|ts_parameter| {
+            if let Ok(binding) = ts_parameter
+                .formal_parameter()
+                .ok()?
+                .as_js_formal_parameter()?
+                .binding()
+            {
+                let syntax = binding.syntax();
+                let text_range = syntax.text_trimmed_range();
+                let text = syntax.text_trimmed().into_text();
+
+                return Some((text_range, text, ts_parameter.syntax().clone()));
+            }
+            None
+        })
+        .collect();
+
+    Some(params)
+}
+
+fn get_all_property_reassignments(members: &JsClassMemberList) -> Vec<AnyJsName> {
+    members
+        .iter()
+        .filter_map(|member| match member {
+            AnyJsClassMember::JsMethodClassMember(_) | AnyJsClassMember::JsSetterClassMember(_) => {
+                Some(
+                    get_assignments_alike_in_node(member.syntax())
+                        .into_iter()
+                        .filter(|operand| contains_this_or_static_member_kind(operand.syntax()))
+                        .filter_map(|operand| {
+                            operand
+                                .as_js_static_member_assignment()
+                                .and_then(|assignment| assignment.member().ok())
+                        }),
+                )
+            }
+            _ => None,
+        })
+        .flatten()
+        .collect()
+}
+
+fn contains_this_or_static_member_kind(assignment: &SyntaxNode<JsLanguage>) -> bool {
+    assignment
+        .descendants_with_tokens(Direction::Next)
+        .any(|node| {
+            matches!(
+                node.kind(),
+                JsSyntaxKind::JS_THIS_EXPRESSION | JsSyntaxKind::JS_STATIC_MEMBER_EXPRESSION
+            )
+        })
+}
```

Comment:
This is the result of using a `SyntaxNode`, which isn't optimal.

Instead create a new node like this

```rs
declare_node_union! {
	pub AnyThisMemberLike = JsThisExpression | JsStaticMemberExpression
}
```

This will be an `enum` that contains all `AstNode` methods and you can query using strongly typed nodes. You will avoid `descendants_with_tokens`, which is expensive